iml-services/iml-device/src/virtual_device/util.rs

```diff
@@ -0,0 +1,153 @@
+        Device::MdRaid(dd) => Some(&mut dd.children),
+        Device::Mpath(dd) => Some(&mut dd.children),
+        Device::VolumeGroup(dd) => Some(&mut dd.children),
+        Device::LogicalVolume(dd) => Some(&mut dd.children),
+        Device::Zpool(dd) => Some(&mut dd.children),
+        Device::Dataset(_) => None,
+    }
+}
+
+pub fn children(d: &Device) -> Option<&OrdSet<Device>> {
+    match d {
+        Device::Root(dd) => Some(&dd.children),
+        Device::ScsiDevice(dd) => Some(&dd.children),
+        Device::Partition(dd) => Some(&dd.children),
+        Device::MdRaid(dd) => Some(&dd.children),
+        Device::Mpath(dd) => Some(&dd.children),
+        Device::VolumeGroup(dd) => Some(&dd.children),
+        Device::LogicalVolume(dd) => Some(&dd.children),
+        Device::Zpool(dd) => Some(&dd.children),
+        Device::Dataset(_) => None,
+    }
+}
+
+pub fn compare_selected_fields(a: &Device, b: &Device) -> bool {
+    selected_fields(a) == selected_fields(b)
+}
+
+fn selected_fields(d: &Device) -> Device {
+    match d {
+        Device::Root(d) => Device::Root(Root {
+            children: OrdSet::new(),
+            ..d.clone()
+        }),
+        Device::ScsiDevice(d) => Device::ScsiDevice(ScsiDevice {
+            children: OrdSet::new(),
+            major: String::new(),
+            minor: String::new(),
+            devpath: PathBuf::new(),
+            paths: OrdSet::new(),
+            ..d.clone()
+        }),
+        Device::Partition(d) => Device::Partition(Partition {
+            children: OrdSet::new(),
+            major: String::new(),
+            minor: String::new(),
+            devpath: PathBuf::new(),
+            paths: OrdSet::new(),
+            ..d.clone()
+        }),
+        Device::MdRaid(d) => Device::MdRaid(MdRaid {
+            children: OrdSet::new(),
+            major: String::new(),
+            minor: String::new(),
+            paths: OrdSet::new(),
+            ..d.clone()
+        }),
+        Device::Mpath(d) => Device::Mpath(Mpath {
+            children: OrdSet::new(),
+            major: String::new(),
+            minor: String::new(),
+            devpath: PathBuf::new(),
+            paths: OrdSet::new(),
+            ..d.clone()
+        }),
+        Device::VolumeGroup(d) => Device::VolumeGroup(VolumeGroup {
+            children: OrdSet::new(),
+            ..d.clone()
+        }),
+        Device::LogicalVolume(d) => Device::LogicalVolume(LogicalVolume {
+            children: OrdSet::new(),
+            major: String::new(),
+            minor: String::new(),
+            devpath: PathBuf::new(),
+            paths: OrdSet::new(),
+            ..d.clone()
+        }),
+        Device::Zpool(d) => Device::Zpool(Zpool {
+            children: OrdSet::new(),
+            ..d.clone()
+        }),
+        Device::Dataset(d) => Device::Dataset(d.clone()),
+    }
+}
+
+pub fn is_virtual(d: &Device) -> bool {
+    match d {
+        Device::Dataset(_)
```

Comment:
`Dataset` and `LogicalVolume` are always children of a virtual device. We should not need to care about iterating to this level and instead be able to lift out the subtree when we hit their single parent.